toolbox/src/DotJNu.cpp

```diff
@@ -4,181 +4,253 @@
+        blockInfo->setInputPortType(INPUT_IDX_JOINT_VEL, PortDataTypeDouble);
 
         if (!success) {
             if (error) error->message = "Failed to configure input ports";
             return false;
         }
 
-        // Output port:
-        // - 6-d vector representing the \dot{J} \dot{q} vector
+        // OUTPUT
+        // ======
+        //
+        // 1) 1x6 Vector representing the \dot{J} \dot{q} vector
+
+        // Number of outputs
         if (!blockInfo->setNumberOfOuputPorts(1)) {
             if (error) error->message = "Failed to configure the number of output ports";
             return false;
         }
 
-        success = blockInfo->setOutputPortVectorSize(0, 6);
-        blockInfo->setOutputPortType(0, PortDataTypeDouble);
+        // Size and type
+        success = blockInfo->setOutputPortVectorSize(OUTPUT_IDX_DOTJ_NU, 6);
+        blockInfo->setOutputPortType(OUTPUT_IDX_DOTJ_NU, PortDataTypeDouble);
 
         return success;
     }
 
-    bool DotJNu::initialize(BlockInformation *blockInfo, wbt::Error *error)
+    bool DotJNu::initialize(BlockInformation* blockInfo, wbt::Error* error)
     {
-        using namespace yarp::os;
+        using namespace iDynTree;
+
         if (!WBIModelBlock::initialize(blockInfo, error)) return false;
 
-        int parentParameters = WBIBlock::numberOfParameters() + 1;
+        // INPUT PARAMETERS
+        // ================
+
         std::string frame;
-        if (!blockInfo->getStringParameterAtIndex(parentParameters, frame)) {
+        int parentParameters = WBIBlock::numberOfParameters();
+
+        if (!blockInfo->getStringParameterAtIndex(parentParameters + 1, frame)) {
             if (error) error->message = "Cannot retrieve string from frame parameter";
             return false;
         }
-        //here obtain joint list and get the frame
-        wbi::iWholeBodyModel * const interface = WBInterface::sharedInstance().model();
-        if (!interface) {
-            if (error) error->message = "Cannot retrieve handle to WBI interface";
+
+        // Check if the frame is valid // TODO: move this somewhere else? WBIBlock?
+        // ---------------------------
+
+        KinDynComputations* const model = WBInterface::sharedInstance().model();
+        if (!model) {
+            if (error) error->message = "Cannot retrieve handle to KinDynComputations";
             return false;
         }
-        wbi::IDList frames =  interface->getFrameList();
+
         if (frame != "com") {
-            if (!frames.idToIndex(wbi::ID(frame), m_frameIndex)) {
+            m_frameIndex = model->getFrameIndex(frame);
+            if (m_frameIndex == FRAME_INVALID_INDEX) {
                 if (error) error->message = "Cannot find " + frame + " in the frame list";
                 return false;
             }
-        } else {
-            m_frameIndex = wbi::wholeBodyInterface::COM_LINK_ID;
         }
-        
+        else {
+            m_frameIndex = COM_LINK_ID;
```

Comment:
@traversaro In the `output()` method of all the blocks that use `COM_LINK_ID` the correct iDynTree's method for CoM is called. Since all integers `<0` are treated as FRAME_INVALID_INDEX by iDynTree, do you have any suggestion on how to handle this case with a viarable?

I mean, when `m_frameIndex = -1` and the frame is `com`, an error is thrown. If `m_frameIndex = -1` and the frame is not `com`, the block goes further to the `output()` and the `COM_LINK_ID` is selected [to chose the alternative method](https://github.com/robotology/WB-Toolbox/pull/56/files/82b48feb52ef097a71c53029e8765ba227b61795#diff-ff0898c2cbf4dce4de854b3e206599bdR241) for computations on CoM.

Reply:
I see, I missed the check on `COM_LINK_ID` . Everything should work fine indeed, but personally I would prefer to have a separated boolean variable to save the information "is this block computing COM-related stuff", rather than overloading the meaning of `m_frameIndex` . However, this is mostly a personal taste, so feel free to use the pattern that you prefer. The only advice that I would recommend is to define  `COM_LINK_ID` to any negative number different from `-1` , as `-1` is exactly the value of `iDynTree::FRAME_INVALID_INDEX`.

Reply:
I like the boolean variable idea. It is the most safe and clean in the long run